scene/gui/text_edit.h

```diff
@@ -452,18 +450,23 @@ class TextEdit : public Control {
 
 	bool selection_drag_attempt = false;
 	bool dragging_selection = false;
+	int drag_and_drop_origin_caret_index = -1;
+	int drag_caret_index = -1;
 
 	Timer *click_select_held = nullptr;
 	uint64_t last_dblclk = 0;
 	Vector2 last_dblclk_pos;
+
+	void _selection_changed(int p_caret = -1);
 	void _click_selection_held();
 
-	void _update_selection_mode_pointer();
-	void _update_selection_mode_word();
-	void _update_selection_mode_line();
+	void _update_selection_mode_pointer(bool p_initial = false);
+	void _update_selection_mode_word(bool p_initial = false);
+	void _update_selection_mode_line(bool p_initial = false);
 
 	void _pre_shift_selection(int p_caret);
-	void _post_shift_selection(int p_caret);
+
+	bool _selection_contains(int p_caret, int p_line, int p_column, bool p_include_edges = true, bool p_only_selections = true) const;
```

Comment:
I could invert it to `p_include_non_selections`? I think `p_only_selections` sounds more clear, though.

Reply:
Thinking about this with the other comment around `mouse_over_selection_caret`, wondering if we should merge them into a single method?

Reply:
I can give `get_selection_at_line_column()` (mouse_over_selection_caret) a `p_only_selections` argument and then it can be used instead of looping in `gui_input()` and `add_caret()`, but I can't remove all uses of `_selection_contains()`, since it is used in `add_caret_at_carets()`.

Reply:
For `add_caret_at_carets` would that be due to the them overlapping?
Otherwise that sounds fine to me.

Reply:
Yes, it's to check if they overlap, so that the correct caret will be removed when they merge.